pydantic/generics.py

```diff
@@ -57,10 +61,10 @@ def __class_getitem__(cls: Type[GenericModelT], params: Union[Type[Any], Tuple[T
             ),
         )
         created_model.Config = cls.Config
-        concrete = all(not _is_typevar(v) for v in concrete_type_hints.values())
+        concrete = all(not _has_typevar(v) for v in concrete_type_hints.values())
         created_model.__concrete__ = concrete
         if not concrete:
-            parameters = tuple(v for v in concrete_type_hints.values() if _is_typevar(v))
+            parameters = tuple(v for v in concrete_type_hints.values() if _has_typevar(v))
             parameters = tuple({k: None for k in parameters}.keys())  # get unique params while maintaining order
```

Comment:
```suggestion
            parameters = tuple({v: None for v in typevars_map.values() if isinstance(v, TypeVar)})
```
Turns out that we were complicating this step more than necessary. We don't need `_has_typevar` at all.

Reply:
Argh... Turns out I was wrong, This fails if one of the params is generic.

Reply:
OK, I worked it out
```suggestion
            parameters = extract_parameters(typevars_map)
```
Where
```python
def extract_parameters(typevars_map: Dict[Any, Any]) -> Tuple[TypeVarType, ...]:
    extracted_parameters = {param: None for param in generate_parameters(list(typevars_map.values()))}
    return tuple(extracted_parameters)


def generate_parameters(v: Any) -> Iterator[TypeVarType]:
    if isinstance(v, TypeVar):
        yield v
    elif lenient_issubclass(v, GenericModel):
        yield from v.__parameters__
    elif isinstance(v, list):
        for var in v:
            yield from generate_parameters(var)
    else:
        args = _get_args(v)
        for arg in args:
            yield from generate_parameters(arg)
```